File: hashmap/word-pattern/word-pattern.ts
function wordPattern(pattern: string, s: string): boolean {
    var words = s.split(' ');

    if (pattern.length != words.length) return false;

    var p2w = new Map<string, string>();
    var w2p = new Map<string, string>();

    for (let i = 0; i < pattern.length; i++)
    {
        const p = pattern[i];
        const w = words[i];

        if (p2w.has(p))
        {
            if (p2w.get(p) != w) return false;
        }
        else
        {
            if (w2p.has(w) && w2p.get(w) != p) return false;
            p2w.set(p, w);
            w2p.set(w, p);
        }
    }

    return true;
};

File: hashmap/word-pattern/word-pattern.cs
public class Solution
{
    public bool WordPattern(string pattern, string s)
    {
        var words = s.Split(' ');

        if (pattern.Length != words.Length) return false;

        var p2w = new Dictionary<char, string>();
        var w2p = new Dictionary<string, char>();

        for (int i = 0; i < pattern.Length; i++)
        {
            char p = pattern[i];
            string w = words[i];

            if (p2w.TryGetValue(p, out var mappedW))
            {
                if (mappedW != w) return false;
            }
            else
            {
                if (w2p.TryGetValue(w, out var mappedP) && mappedP != p) return false;
                p2w[p] = w;
                w2p[w] = p;
            }
        }

        return true;
    }
}